street_network/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,128 @@
+    for s in 0..road_ids.len() {
+        for d in 0..road_ids.len() {
+            if s == d {
+                continue;
+            }
+            // FIXME check if we can travel into the intersection on s and out of the intersection on d
+            if turn_is_allowed(roads.get(s).unwrap(), road_ids.get(d).unwrap()) {
+                connections.push((s, d));
             }
         }
+    }
 
-        if num_roads_inout == 0 {
-            // The simplest MultiConnect
-            return MultiConnection;
+    // Calculate all the collisions.
+    let mut conflicts = BTreeMap::new();
+    let mut worst_conflict = Uncontested;
+    // Compare every pair of connections. Use the order of the roads around the intersection to
+    // detect if they diverge, merge, or cross.
+    // assert!(connections is sorted) so small_con large_con makes sense.
+    let mut each_con = connections.iter();
+    while let Some(small_con) = each_con.next() {
+        for large_con in each_con.clone() {
+            let conflict = calc_conflict(small_con, large_con);
+            worst_conflict = max(worst_conflict, conflict);
+            conflicts.insert((small_con, large_con), conflict);
         }
+    }
+
+    match worst_conflict {
+        Cross => (Crossing, Cross),
+        c => (MultiConnection, c),
+    }
+}
 
-        // detect the simple case of a single road splitting
-        if num_roads_inout == 1 {
-            // TODO Determine if it is possible to turn from the 'in' road to the 'out' road.
-            // If not, then we have a dual carriageway split.
-            if false {
-                return MultiConnection;
+fn turn_is_allowed(src: &Road, dst_id: &OriginalRoad) -> bool {
+    let mut has_exclusive_allows = false;
+    for (t, other) in &src.turn_restrictions {
+        match t {
+            RestrictionType::BanTurns => {
+                if other == dst_id {
+                    return false;
+                }
+            }
+            RestrictionType::OnlyAllowTurns => {
+                if other == dst_id {
+                    return true;
+                }
+                has_exclusive_allows = true;
             }
         }
     }
+    !has_exclusive_allows
+}
 
-    // A Merge is characterised by an uninterrupted road (that would qualify as a Connection or
-    // MultiConnection), with additional connected roads that yield.
-    // TODO Combine roads into corridors and count them.
+fn calc_conflict(a: &(usize, usize), b: &(usize, usize)) -> ConflictType {
+    use ConflictType::*;
+    if a.0 == b.0 {
+        return Diverge;
+    }
+    if a.1 == b.1 {
+        return Merge;
+    }
+    if a.0 == b.1 || a.1 == b.0 {
+        // TODO depends on driving side and arm order?
+    }
+    if is_between(a.0, b) ^ is_between(a.1, b) {
+        return Cross;
+    }
+    return Uncontested;
+}
 
-    return Crossing;
+fn is_between(num: usize, range: &(usize, usize)) -> bool {
+    let bot = min(range.0, range.1);
```

Comment:
We're reasoning about a cyclic group here, since it's roads sorted clockwise and there's an arbitrary start/end. IIUC, the simple case is range.0 < num < range.1... if range.0 is < range.1. Argh actually I just need to draw this to understand. Are you confident the logic here is correct? :P  Maybe we can add unit tests if we're unsure

Reply:
I've taken the cyclic group and one connection. The group is like a circle with the connection running through the middle joining a pairs of points. Cut the circle somewhere to arbitrarily define the start/end of our reference list, i.e. where zero is. Note that any point on the circle is either between the pair of points, or between one point and the start/end. For an additional connection to cross the existing connection, one end would have to be between the pair, and one outside. Hence `if is_between(a.0, b) ^ is_between(a.1, b)`. `is_between` is over ordinary numbers.

I don't know if I followed it through accurately, and seeing it all come together under tests will be one way to find out, but that was my idea.

Reply:
![IMG_20220922_133951.jpg](https://user-images.githubusercontent.com/3784591/191666819-9bfbfaaf-f507-4f46-a586-e5a87348420c.jpg)

Zero at `O`, connection `A` to `B`. Consider starting a comnection from `X` or `Y`

Reply:
Thank you! I understand the logic better now and think this is correct... but yeah, end-to-end tests will be the easiest to flush out problems anywhere